apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright 2017, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.using System;
+
+using Google.Api.Gax;
+using Google.Cloud.Firestore.V1Beta1;
+using System;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Firestore.Data
+{
+    // TODO: Abstract with all the trimmings?
+
+    /// <summary>
+    /// A Firestore database
+    /// </summary>
+    public sealed class FirestoreDb
+    {
+        private const string DefaultDatabaseId = "(default)";
+
+        /// <summary>
+        /// The underlying client used to make service requests.
+        /// </summary>
+        public FirestoreClient Client { get; }
+
+        /// <summary>
+        /// The project ID associated with this database.
+        /// </summary>
+        public string ProjectId { get; }
+
+        /// <summary>
+        /// The database ID associated with this database.
+        /// </summary>
+        public string DatabaseId { get; }
+
+        /// <summary>
+        /// The resource name of the database, in the form "projects/{project_id}/databases/{database_id}".
+        /// </summary>
+        internal string RootPath { get; }
+
+        /// <summary>
+        /// The root of the documents, in the form "projects/{project_id}/databases/{database_id}/documents".
+        /// </summary>
+        internal string DocumentsPath { get; }
+
+        private FirestoreDb(string projectId, string databaseId, FirestoreClient client)
+        {
+            ProjectId = GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+            DatabaseId = GaxPreconditions.CheckNotNull(databaseId, nameof(databaseId));
+            Client = GaxPreconditions.CheckNotNull(client, nameof(client));
+            RootPath = $"projects/{ProjectId}/databases/{DatabaseId}";
```

Comment:
Is there a code-gen'ed ResourceName that could be used?

Reply:
There's DatabaseName for this part, but I don't think we get beyond that - we have DocumentPathName, but I don't think it's useful. I think at the moment I'll stick to just strings, but I'll add a TODO to investigate. (I don't think it'll affect the public API surface.)